### DISCUSSION TITLE
Question about example Relative Permeability Estimation

### DISCUSSION THREAD

**Discussion by qingqi-code** [PMEAL/OpenPNM#2050] [Q&A]
Hi, I am trying to study the code about relative permeability. When I run the example, I got an AttributeError.
When I try to run:
```python
import numpy as np
import pandas as pd
import openpnm as op
%config InlineBackend.figure_formats = ['svg']
import openpnm.models.physics as pmods
import matplotlib.pyplot as plt
pn = op.network.Cubic(shape=[15,15,15], spacing=6e-5)
geom = op.geometry.StickAndBall(network=pn, pores=pn['pore.all'],
                                throats=pn['throat.all'])
air = op.phases.Air(network=pn,name='air')
water = op.phases.Water(network=pn,name='water')
air.add_model(propname='throat.hydraulic_conductance',
              model=pmods.hydraulic_conductance.hagen_poiseuille)
air.add_model(propname='throat.entry_pressure',
              model=pmods.capillary_pressure.washburn)
water.add_model(propname='throat.hydraulic_conductance',
                model=pmods.hydraulic_conductance.hagen_poiseuille)
water.add_model(propname='throat.entry_pressure',
                model=pmods.capillary_pressure.washburn)
ip = op.algorithms.InvasionPercolation(network=pn, phase=air)
Finlets= pn.pores('top')
ip.set_inlets(pores=Finlets)
ip.run()
air.update(ip.results())
rp = op.algorithms.metrics.RelativePermeability(network=pn)
rp.settings.update({'nwp': 'air',
                    'invasion_sequence': 'invasion_sequence'})
rp.run(Snwp_num=10)
```
it comes
```
AttributeError                            Traceback (most recent call last)
<ipython-input-23-0187d3b3286c> in <module>
      2 rp.settings.update({'nwp': 'air',
      3                     'invasion_sequence': 'invasion_sequence'})
----> 4 rp.run(Snwp_num=10)

~\anaconda3\lib\site-packages\openpnm\algorithms\metrics\RelativePermeability.py in run(self, Snwp_num)
    322             Snwp_num = self.settings['Snwp_num']
    323         net = self.project.network
--> 324         K_dir = set(self.settings['flow_inlets'].keys())
    325         for dim in K_dir:
    326             flow_pores = [net.pores(self.settings['flow_inlets'][dim]),

AttributeError: 'NoneType' object has no attribute 'keys'
```
So I think it asks me to specify the  the flow_inlet? If so how can I set this value? Thank you so much!

**Comment by Ni2M** [comment]
Hi, Since the last release (Sept 8) we have changed the RelativePermeability.py on [dev branch.](https://github.com/PMEAL/OpenPNM/blob/dev/openpnm/algorithms/metrics/RelativePermeability.py)  and its example, accordingly. The code works fine on my computer and I am using the latest dev branch that is compatible with the example. This branch will be released next time. I believe you are using the last release of OpenPNM (or release branch) on which these changes have not been applied yet.  I would suggest one of the following options:
1) You can continue with your installed OpenPNM, but instead of using that relperm example (provided above) please use the relperm example that is associated with the release version. It can be found on [release branch on github](https://github.com/PMEAL/OpenPNM/blob/release/examples/notebooks/algorithms/multiphase/relative_permeability_metrics.ipynb)
The released branch example can also be found in your OpenPNM's installed folder. Usually can be found on C:\Usersname\anaconda3\lib\site-packages\openpnm\examples\notebooks\algorithms\multiphase\relative_permeability_metrics.ipynb.
You will notice that this file for example has ```rp.setup(invading_phase='air',invasion_sequence='invasion_sequence')``` instead of the```rp.settings.update({'nwp': 'air', 'invasion_sequence': 'invasion_sequence'}```.
2) You can alternatively: Clone OpenPNM using Git and install it using the directory of your OpenPNM folder (refere to [developers](https://github.com/PMEAL/OpenPNM/tree/dev#for-developers)). Then checkout/pull the "dev" branch on Git. Now that you are using dev branch, that example code (provided above) should work this time.
However, the changes on current dev branch will be released on our next OpenPNM version.

**Reply by qingqi-code** [reply]
Thank you so much. I just followed the tutorial and did not watch the dev branch. So it works now. And may I ask another question. Will the wetting condition (aka contact angle) be used in calculating relative permeability? And can I apply heterogeneous wetting condition? I guess I can assign a list of specified distribution of contact angle to water phase/air phase?

**Reply by jgostick** [reply]
Hi @qingqi-code 
Changing contact angle from A to B will not change the rel perm curves since it will not alter the invasion pattern.  If you assigned a distribution of contact angles however, then yes things may change a bit...but it will be subtle.  Also note that if you plan to do 'imbibition' you must use the mixed percolation algorithm and it's quite a bit more complicated than the standard invasion percolation.

**Reply by qingqi-code** [reply]
Thank you very much Dr. Gostick. So can I use ['MixedInvasionPercolationCoop.py'](https://github.com/PMEAL/OpenPNM/blob/4ba817251e7ce4d777ec34605a991123c01bd8cf/openpnm/algorithms/MixedInvasionPercolationCoop.py) as percolation algorithm if I want to focus on the heterogeneous contact angle influence on relative permeability? Or I need to custom an algorithm by myself.  Thank you so much!

**Reply by jgostick** [reply]
If your invading phase is also the non-wetting phase then you can use normal invasion percolation.  If your invading the wetting phase, then you need to use mixed percolation.  If your invading phase is BOTH wetting and non-wetting, like the contact angle varies throughout the medium from 40 to 140 for eg, then you'll need to use mixed percolation.  There is a pretty good example for you to look at [here](https://github.com/PMEAL/OpenPNM/blob/dev/examples/notebooks/algorithms/percolation/C_mixed_invasion_percolation.ipynb)

**Reply by qingqi-code** [reply]
Thank you for your help professor! I will use it.
